orsopy/fileio/base.py

```diff
@@ -45,93 +44,71 @@ def _noop(self, *args, **kw):
-                seen_default = True
-            elif seen_default:
-                raise TypeError(f"non-default argument {f.name!r} " "follows default argument")
-
-    locals = {f"_type_{f.name}": f.type for f in fieldsarg}
-    locals.update({"MISSING": MISSING, "_HAS_DEFAULT_FACTORY": _HAS_DEFAULT_FACTORY})
-
-    body_lines = []
-    for f in fieldsarg:
-        line = _field_init(f, frozen, locals, self_name)
-        if line:
-            body_lines.append(line)
-
-    if has_post_init:
-        params_str = ",".join(f.name for f in fieldsarg if f._field_type is _FIELD_INITVAR)
-        body_lines.append(f"{self_name}.{_POST_INIT_NAME}({params_str})")
-
-    # processing of additional user keyword arguments
-    body_lines += ["for k,v in user_kwds.items():", "    setattr(self, k, v)"]
-
-    return _create_fn(
-        "__init__",
-        [self_name] + [_init_param(f) for f in fieldsarg if f.init] + ["**user_kwds"],
-        body_lines,
-        locals=locals,
-        globals=globals,
-        return_type=None,
-    )
-
-
-# register all ORSO classes here:
-ORSO_DATACLASSES = dict()
-
-
-def orsodataclass(cls: type):
-    ORSO_DATACLASSES[cls.__name__] = cls
-    attrs = cls.__dict__
-    bases = cls.__bases__
-    if "__annotations__" in attrs and len([k for k in attrs["__annotations__"].keys() if not k.startswith("_")]) > 0:
-        # only applies to dataclass children of Header
-        # add optional comment attribute, needs to come last
-        attrs["__annotations__"]["comment"] = Optional[str]
-        setattr(cls, "comment", field(default=None))
-
-        # create the _orso_optional attribute
-        orso_optionals = []
-        for fname, ftype in attrs["__annotations__"].items():
-            if type(None) in get_args(ftype):
-                orso_optionals.append(fname)
-        for base in bases:
-            if hasattr(base, "_orso_optionals"):
-                orso_optionals += getattr(base, "_orso_optionals")
-        setattr(cls, "_orso_optionals", orso_optionals)
-        out = dataclass(cls, repr=False, init=False)
-        fieldsarg = getattr(out, _FIELDS)
-
-        # Generate custom __init__ method that allows arbitrary extra keyword arguments
-        has_post_init = hasattr(out, _POST_INIT_NAME)
-        # Include InitVars and regular fields (so, not ClassVars).
-        flds = [f for f in fieldsarg.values() if f._field_type in (_FIELD, _FIELD_INITVAR)]
-        init_fun = _custom_init_fn(flds, False, has_post_init, "self", globals())
-        _set_new_attribute(out, "__init__", init_fun)
-
-        return out
-    else:
-        return cls
+class ORSOResolveError(ValueError):
+    pass
 
 
-class ORSOResolveError(ValueError):
+class ORSOSchemaWarning(RuntimeWarning):
     pass
 
 
 class Header:
     """
-    The super class for all of the items in the orso module.
+    The super class for all the items in the orso module.
     """
 
     _orso_optionals: List[str] = []
+    _subclass_dict_ = {}
```

Comment:
This dictionary is actually not needed, was just a convenience to put it in the class.

What I didn't like so much about the previous solution was the use of a global variable to track something that is associated with a class and to search through that list instead of using the isinstance method. I thought this would be more clear to see what it does. But we can discuss that.